src/state/psarstake/hooks.ts

```diff
@@ -0,0 +1,584 @@
+  }, []);
+
+  const handleMax = useCallback(() => {
+    maxAmountInput && onUserInput(maxAmountInput.toExact());
+    setStepIndex(4);
+  }, [maxAmountInput, onUserInput]);
+
+  const onChangePercentage = (value: number) => {
+    if (stakedAmount.lessThan('0')) {
+      setTypedValue('0');
+      return;
+    }
+    if (value === 100) {
+      setTypedValue(stakedAmount.toExact());
+    } else if (value === 0) {
+      setTypedValue('0');
+    } else {
+      const newAmount = stakedAmount.multiply(JSBI.BigInt(value)).divide(JSBI.BigInt(100)) as TokenAmount;
+
+      setTypedValue(newAmount.toSignificant(6));
+    }
+  };
+
+  const onUnstake = async () => {
+    if (!sarStakingContract || !parsedAmount || !position) {
+      return;
+    }
+    setAttempting(true);
+    try {
+      const estimatedGas = await sarStakingContract.estimateGas.withdraw(
+        position.id.toHexString(),
+        `0x${parsedAmount.raw.toString(16)}`,
+      );
+      const response: TransactionResponse = await sarStakingContract.withdraw(
+        position.id.toHexString(),
+        `0x${parsedAmount.raw.toString(16)}`,
+        {
+          gasLimit: calculateGasMargin(estimatedGas),
+        },
+      );
+      await waitForTransaction(library, response, 5);
+      addTransaction(response, {
+        summary: t('sarUnstake.transactionSummary', { symbol: png.symbol, balance: parsedAmount.toSignificant(2) }),
+      });
+      setHash(response.hash);
+    } catch (error) {
+      const err = error as any;
+      if (err?.code !== 4001) {
+        console.error(err);
+        setUnstakeError(err?.message);
+      }
+    } finally {
+      setAttempting(false);
+    }
+  };
+
+  return useMemo(
+    () => ({
+      attempting,
+      hash,
+      stepIndex,
+      typedValue,
+      parsedAmount,
+      error,
+      unstakeError,
+      onUserInput,
+      wrappedOnDismiss,
+      handleMax,
+      onUnstake,
+      onChangePercentage,
+      setStepIndex,
+    }),
+    [attempting, typedValue, parsedAmount, hash, stepIndex, error, account, onUserInput, handleMax, position],
+  );
+}
+
+export function useDerivativeSarCompound(position: Position | null) {
+  const [attempting, setAttempting] = useState(false);
+  const [hash, setHash] = useState<string | null>(null);
+  const [compoundError, setCompoundError] = useState<string | null>(null);
+
+  const { account } = usePangolinWeb3();
+  const { library } = useLibrary();
+
+  const sarStakingContract = useSarStakingContract();
+
+  const { t } = useTranslation();
+  const addTransaction = useTransactionAdder();
+
+  const wrappedOnDismiss = useCallback(() => {
+    setCompoundError(null);
+    setHash(null);
+    setAttempting(false);
+  }, []);
+
+  const onCompound = async () => {
+    if (!sarStakingContract || !position) {
+      return;
+    }
+    setAttempting(true);
+    try {
+      const estimatedGas = await sarStakingContract.estimateGas.compound(position.id.toHexString());
```

Comment:
just out of curiosity, is it compulsory to estimate gas here ?

Reply:
Maybe it's not, but I put it to a form of concern